### PR TITLE
Fix FEEvaluation::dofs_per_cell for vector-valued case

### LINKED ISSUE #5343: Fix FEEvaluation::dofs_per_cell for vector-valued case
Due to some mistakes back in 2012 we set `FEEvaluation::dofs_per_cell` as the number of degrees of freedom on a scalar component also for vector-valued problems. Even though this might cause some trouble in downstream projects, we should fix this because it is very confusing.

### PR DESCRIPTION
**Pull request by kronbichler**
Fixes #5343.

When fixing this member variable, I also added another one for `dofs_per_component` and `static_dofs_per_cell` that equals `tensor_dofs_per_cell` to reflect the analogy to `static_n_q_points`.

### CODE REVIEW

File: doc/news/changes/minor/20171028MartinKronbichler [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Fixed: The member variable FEEvaluation::dofs_per_cell and the static variable
```
Comment (davydden) [inline doc/news/changes/minor/20171028MartinKronbichler]: maybe should go into incompatibilities?

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 2]
```diff
@@ -1910,7 +1913,9 @@ class FEEvaluation : public FEEvaluationAccess<dim,n_components_,Number>
   static const unsigned int dimension     = dim;
   static const unsigned int n_components  = n_components_;
   static const unsigned int static_n_q_points    = Utilities::fixed_int_power<n_q_points_1d,dim>::value;
-  static const unsigned int tensor_dofs_per_cell = Utilities::fixed_int_power<fe_degree+1,dim>::value;
+  static const unsigned int static_dofs_per_component = Utilities::fixed_int_power<fe_degree+1,dim>::value;
+  static const unsigned int tensor_dofs_per_cell = n_components *static_dofs_per_component;
```
Comment (davydden) [inline include/deal.II/matrix_free/fe_evaluation.h]: maybe add space after `*`?

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 3]
```diff
@@ -2024,16 +2029,26 @@ class FEEvaluation : public FEEvaluationAccess<dim,n_components_,Number>
...
+   * the underlying evaluation object. Usually close to
+   * static_dofs_per_component, but the number depends on the actual element
+   * selected and is thus not static.
+   */
+  const unsigned int dofs_per_component;
+
+  /**
+   * The number of degrees of freedom on the cell accumulated over all
+   * components in the current evaluation object. Usually close to
+   * static_dofs_per_component*n_components=static_dofs_per_cell, but the
```
Comment (davydden) [inline include/deal.II/matrix_free/fe_evaluation.h]: maybe `static_dofs_per_cell=static_dofs_per_component*n_components` to ease reading?

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 4]
```diff
@@ -5084,7 +5099,8 @@ ::FEEvaluation (const MatrixFree<dim,Number> &data_in,
                 const unsigned int quad_no)
   :
   BaseClass (data_in, fe_no, quad_no, fe_degree, static_n_q_points),
-  dofs_per_cell (this->data->dofs_per_cell),
+  dofs_per_component (this->data->dofs_per_component_on_cell),
+  dofs_per_cell (this->data->dofs_per_component_on_cell *n_components_),
```
Comment (davydden) [inline include/deal.II/matrix_free/fe_evaluation.h]: space after `*`? Same below.

File: doc/news/changes/minor/20171028MartinKronbichler (hunk 1 above)
Comment (kronbichler) [reply doc/news/changes/minor/20171028MartinKronbichler]: I thought about that too - but then decided it equally valid to say that just fixes up what the variable name says. I have no strong opinion, so we could change it.

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 2 above)
Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h]: That was `astyle`, no idea why it triggers here...

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: ah, ok then.

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 4 above)
Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h]: `astyle` here as well...

File: doc/news/changes/minor/20171028MartinKronbichler (hunk 1 above)
Comment (davydden) [reply doc/news/changes/minor/20171028MartinKronbichler]: i don't have a strong opinion either

**Comment by drwells** [comment]
I believe that you three are the biggest users of the MatrixFree frame work, so if you three approve then things should be right. 👍

**Pull request merged**